src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,268 @@
+  }
+
+  // return the permissions object
+  public getPermissions() {
+    const permissions = this.permissions;
+    // reset permissions
+    this.permissions = {};
+
+    return permissions;
+  }
+
+  /*
+    generate query DSL by the specific conditions, used for fetching saved objects from the saved objects index
+    */
+  public genereateGetPermittedSavedObjectsQueryDSL(
+    permissionType: string,
+    savedObjectType?: string | string[],
+    user?: string,
+    group?: string
+  ) {
+    if ((!user && !group) || !permissionType) {
+      return {
+        query: {
+          match_none: {},
+        },
+      };
+    }
+
+    const bool: any = {
+      filter: [],
+    };
+    const subBool: any = {
+      should: [],
+    };
+    if (!!user) {
+      subBool.should.push({
+        term: {
+          ['permissions.' + permissionType + '.users']: user,
+        },
+      });
+      subBool.should.push({
+        term: {
+          ['permissions.' + permissionType + '.users']: '*',
+        },
+      });
+    }
+    if (!!group) {
+      subBool.should.push({
+        term: {
+          ['permissions.' + permissionType + '.groups']: group,
+        },
+      });
+      subBool.should.push({
+        term: {
+          ['permissions.' + permissionType + '.groups']: '*',
+        },
+      });
+    }
+
+    bool.filter.push({
+      bool: subBool,
+    });
+
+    if (!!savedObjectType) {
+      bool.filter.push({
+        terms: {
+          type: Array.isArray(savedObjectType) ? savedObjectType : [savedObjectType],
+        },
+      });
+    }
+
+    return { query: { bool } };
+  }
+
+  /* 
+    transfrom permissions format
+    input:   {
+        read: {
+            users:['user1']
+        },
+        write:{
+            groups:['group1']
+        }
+    }
+
+    output: [
+        {type:'user',name:'user1',permissions:['read']},
+        {type:'group',name:'group1',permissions:['write']},
+    ]
+    */
+  public transformPermissions(permissions: Permissions) {
+    const result: any = [];
+    if (!permissions) {
+      return result;
+    }
+    const userPermissionMap: Map<string, string[]> = new Map();
+    const groupPermissionMap: Map<string, string[]> = new Map();
+    for (const permissionType in permissions) {
+      if (!!permissionType) {
+        const value = permissions[permissionType];
+        if (value?.users) {
+          for (const user of value?.users) {
+            if (!userPermissionMap.get(user)) {
+              userPermissionMap.set(user, []);
+            }
+            userPermissionMap.set(user, [...userPermissionMap.get(user)!, permissionType]);
+          }
+        }
+        if (value?.groups) {
+          for (const group of value?.groups) {
+            if (!groupPermissionMap.get(group)) {
+              groupPermissionMap.set(group, []);
+            }
+            groupPermissionMap.set(group, [...groupPermissionMap.get(group)!, permissionType]);
+          }
+        }
+      }
+    }
```

Comment:
```suggestion
    const permissionMapResult: Record<string, Record<string, string[]>> = {};
    const principleTypes = ['users', 'groups'];
    for (const permissionType in permissions) {
      if (!!permissionType) {
        const value = permissions[permissionType];
        principleTypes.forEach(principleType => {
          if (value?.[principleType]) {
            for (const principle of value?.[principleType]) {
              if (!permissionMapResult[principleType]) {
                permissionMapResult[principleType] = {};
              }
              if (!permissionMapResult[principleType][principle]) {
                permissionMapResult[principleType][principle] = [];
              }
              permissionMapResult[principleType][principle] = [...permissionMapResult[principleType][principle]!, permissionType];
            }
          }
        });
      }
    }
```

Reply:
I think `principle` should be `principal`?

Reply:
Yes, change that.